### PR TITLE
CRM-20102 - Case Activity api improvements

### PR DESCRIPTION
**Pull request by colemanw**
* [CRM-20102: Case table view](https://issues.civicrm.org/jira/browse/CRM-20102)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
FAIL! :-)

**Comment by colemanw** [comment]
@eileenmcnaughton want to trade one for this?

**Comment by eileenmcnaughton** [comment]
@colemanw this is quite long so I'm going to angle for 2-4-1

#9936 & #9655

File: CRM/Case/BAO/Case.php [hunk 1]
```diff
@@ -1212,9 +1212,20 @@ public static function getCaseActivity($caseID, &$params, $contactID, $context =
    *
    */
   public static function getRelatedContacts($caseID, $skipDetails = FALSE) {
+    $caseRoles = array();
+    if (!$skipDetails) {
+      $caseInfo = civicrm_api('Case', 'getsingle', array(
```
Comment (eileenmcnaughton) [inline CRM/Case/BAO/Case.php]: can we not use civicrm_api3?

There is acceptable reason for an error here so an uncaught exception would be appropriate

Comment (colemanw) [inline CRM/Case/BAO/Case.php]: When I was using `civicrm_api3` it caused a fatal in the syntaxConformanceTests. I'm not exactly sure why, but maybe due to the way they use transactions it was not able to find the case?

Comment (eileenmcnaughton) [inline CRM/Case/BAO/Case.php]: I'll try the test out locally

Comment (seamuslee001) [inline CRM/Case/BAO/Case.php]: Just a thought, potentially maybe the issue the the syntaxConformanceTests may have been that civicrm_api3 throws an exception in error where as civicrm_api returns array('is_error' => TRUE  etc)

File: CRM/Case/BAO/Case.php [hunk 2]
```diff
@@ -1212,9 +1212,20 @@ public static function getCaseActivity($caseID, &$params, $contactID, $context =
    *
    */
   public static function getRelatedContacts($caseID, $skipDetails = FALSE) {
```
Comment (eileenmcnaughton) [inline CRM/Case/BAO/Case.php]: I think I wanna get all @xurizaemon on you & complain about the double negative. Can we do $includeDetails for readability

Comment (eileenmcnaughton) [inline CRM/Case/BAO/Case.php]: Also, I feel the default should be NOT to add additional details. Don't slow down calls unless they want the info as a principle

File: Civi/API/Api3SelectQuery.php [hunk 3]
```diff
@@ -106,7 +106,7 @@ protected function buildWhereClause() {
         continue;
       }
       $operator = is_array($value) ? \CRM_Utils_Array::first(array_keys($value)) : NULL;
-      if (!in_array($operator, \CRM_Core_DAO::acceptedSQLOperators())) {
+      if (!in_array($operator, \CRM_Core_DAO::acceptedSQLOperators(), TRUE)) {
```
Comment (eileenmcnaughton) [inline Civi/API/Api3SelectQuery.php]: OK, so this is changing to strict checking.... can you explain why this change?

File: api/v3/Activity.php [hunk 4]
```diff
@@ -392,18 +392,12 @@ function civicrm_api3_activity_get($params) {
...
   }
 
-  $returns = CRM_Utils_Array::value('return', $params, array());
-  if (!is_array($returns)) {
-    $returns = str_replace(' ', '', $returns);
-    $returns = explode(',', $returns);
-  }
-  $returns = array_fill_keys($returns, 1);
-
+  $returns = $options['return'];
```
Comment (eileenmcnaughton) [inline api/v3/Activity.php]: Yep, better idea!

File: api/v3/Activity.php [hunk 5]
```diff
@@ -482,6 +476,14 @@ function _civicrm_api3_activity_get_formatResult($params, $activities) {
         }
         break;
 
+      case 'case_id':
```
Comment (eileenmcnaughton) [inline api/v3/Activity.php]: Why don't we use the @colemanw magic join to add this onto the query if wanted

File: api/v3/Case.php [hunk 6]
```diff
@@ -218,16 +218,18 @@ function civicrm_api3_case_get($params) {
 
   // Add clause to search by client
   if (!empty($params['contact_id'])) {
-    $contacts = array();
-    foreach ((array) $params['contact_id'] as $c) {
-      if (!CRM_Utils_Rule::positiveInteger($c)) {
-        throw new API_Exception('Invalid parameter: contact_id. Must provide numeric value(s).');
+    // Convert from nonstandard format
```
Comment (eileenmcnaughton) [inline api/v3/Case.php]: I think it would be better to state what the non standard form is & the fact you are converting for legacy compatibility

File: CRM/Case/BAO/Case.php (hunk 1 above)
Comment (eileenmcnaughton) [inline CRM/Case/BAO/Case.php]: Ok I found the problem

_civicrm_api3_case_read 

calls getRelatedContacts causing a code loop & tests to fail. Defaulting to not getting the additional data will fix this (as well as reduce overhead - per comment later)

File: api/v3/Case.php [hunk 7]
```diff
@@ -218,16 +218,18 @@ function civicrm_api3_case_get($params) {
...
-      $contacts[] = $c;
     }
-    $sql
-      ->join('civicrm_case_contact', 'INNER JOIN civicrm_case_contact ON civicrm_case_contact.case_id = a.id')
-      ->where('civicrm_case_contact.contact_id IN (' . implode(',', $contacts) . ')');
+    else {
+      $params['contact_id'] = array('=' => $params['contact_id']);
+    }
+    $clause = CRM_Core_DAO::createSQLFilter('contact_id', $params['contact_id']);
+    $sql->where("a.id IN (SELECT case_id FROM civicrm_case_contact WHERE $clause)");
```
Comment (eileenmcnaughton) [inline api/v3/Case.php]: I'm not loving this table alias of 'a' - is that 'whatever the base table is'

**Comment by eileenmcnaughton** [comment]
Looks good overall - would be good to get a test on 'case_type_id.definition' as well, but you did increase test coverage ... a smidgeon

File: Civi/API/Api3SelectQuery.php (hunk 3 above)
Comment (colemanw) [inline Civi/API/Api3SelectQuery.php]: It's a precaution. As I said in the commit message, there's a bug in php (at least some versions) where the function incorrectly returns `true` if the first param is `0`.

File: api/v3/Activity.php (hunk 5 above)
Comment (colemanw) [inline api/v3/Activity.php]: Believe me I tried. The query function just doesn't want to join across 2 tables like that.
My frustration with this bit was what prompted my comments about joins over in the api4 discussion.

File: api/v3/Case.php (hunk 7 above)
Comment (colemanw) [inline api/v3/Case.php]: Yes that's `\Civi\API\SelectQuery::MAIN_TABLE_ALIAS`

File: CRM/Case/BAO/Case.php (hunk 2 above)
Comment (colemanw) [inline CRM/Case/BAO/Case.php]: I didn't invent this function, but it's only invoked from the api if it's requested in the `return` params.

File: CRM/Case/BAO/Case.php (hunk 1 above)
Comment (colemanw) [inline CRM/Case/BAO/Case.php]: I don't get why there would be an infinite loop. In this call the `return` param is set so there should be no recursion into calling this again.
Also why would `civicrm_api3` trigger a loop but `civicrm_api` would not?

File: api/v3/Activity.php (hunk 4 above)
Comment (colemanw) [inline api/v3/Activity.php]: Yes, unfortunately there's a weird quirk where the `_civicrm_api3_get_options_from_params` function only supports the legacy `return.foo` syntax under certain circumstances so that's why I could take out the above lines but had to leave in the below.

File: Civi/API/Api3SelectQuery.php (hunk 3 above)
Comment (eileenmcnaughton) [inline Civi/API/Api3SelectQuery.php]: ok

File: api/v3/Case.php (hunk 7 above)
Comment (eileenmcnaughton) [inline api/v3/Case.php]: ok

File: api/v3/Activity.php (hunk 5 above)
Comment (eileenmcnaughton) [inline api/v3/Activity.php]: ok

File: CRM/Case/BAO/Case.php (hunk 2 above)
Comment (eileenmcnaughton) [inline CRM/Case/BAO/Case.php]: yep, from a coding POV I think getting additional data should be opt in, not opt out. Reading "IF NOT SKIP DETAIL" is brain hurty & means the same thing as "IF INCLUDE CASEROLEDETAILS" which seems much more readable.

 I guess you could argue the api should opt in to getting roles, but there is a third place in the code that calls this function.

File: CRM/Case/BAO/Case.php (hunk 1 above)
Comment (eileenmcnaughton) [inline CRM/Case/BAO/Case.php]: So, the other thing here is... are you just accessing this via the api? This function is called from 3 places, 1 is the api & I feel like your changes are only meant for the benefit of the api. Obviously you need to change the query in this function - but I feel like the wrangling would be better in the api layer. 

It seems the api has the 'case_type_id.definition', already resolved (& the 2 other places in the code don't care about the extra data & will probably e-notice when they hit this line.

$role = CRM_Utils_Array::value($details['role'], $caseRoles);

As a general rule I support pushing stuff into the BAO layer for create & delete & actions, but doing handling for get actions in the API layer rather than in 'random BAO functions'

I'll have another go at the tests, thinking about it it makes sense that there must be able to quietly fail on an enotice or similar with civicrm_api but it noisily fails with civicrm-api3

**Comment by colemanw** [comment]
@eileenmcnaughton - found the cause - the test was picking up an actual bug, whadda ya know?
Have increased test coverage as penance.

**Comment by eileenmcnaughton** [comment]
@colemanw yay cool, My only issue now is just the use of $skipDetails = FALSE rather than $includeDetails = FALSE. Perhaps I've been listening to @xurizaemon too much but I agree that skip is a code smell

**Comment by colemanw** [comment]
I fixed the double-negative, just for you @eileenmcnaughton.

However there's still one test failure. This one had me scratching my head for a while, finally figured out what's going on:

- The `case_type_id` field is not marked `required` in the schema.
- The `CRM_Core_DAO::createTestObject` function therefore ignores it.
- So the `syntaxConformanceTest` is creating mock cases with no `case_type_id`, which is absurd.
- We then (appropriately) get a hard fail when trying to retrieve these cases using api3 in the `getRelatedContacts` function.

**Comment by eileenmcnaughton** [comment]
There are 3 places that call getRelatedContacts - one you have set to !$includeDetails, the api is defaulting to return & the last one - does that definitely need the extra load?

Re the tests, I think there is a way of declaring fields on an api unworkable in the context of the singleValueAlter test - I'm assuming there is no problem unless case_id is set in the return values.

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by colemanw** [comment]
@eileenmcnaughton making the field `required` in the schema solves the problem. Does that look ok fix to you?

**Comment by eileenmcnaughton** [comment]
let's do it!

**Comment by eileenmcnaughton** [comment]
commits were against the epic rather than a sub-issue (not ideal) - couldn't figure what to close.

**Comment by agh1** [comment]
Could anyone provide a concise explanation of what this does?  The Jira ticket is kind of useless, and every time I skim the code to summarize for the release notes, I find something different that doesn't fit in with the rest.

**Comment by colemanw** [comment]
@agh1 sure. Sorry for not being more descriptive:

- The `Activity.get` api now accepts `case_id` param as a filter or return value.
- The `Case.get` api now properly handles sql operators for the `contact_id` field.
- The `Case.get` api now returns the `case_id.definition` joined field with proper formatting.

**Comment by agh1** [comment]
Awesome--thanks @colemanw!
